### PR TITLE
[WIP] Add summa

### PR DESCRIPTION
**Pull request by bartnijssen**
**Description:** SUMMA (https://github.com/NCAR/summa) is a hydrological model written in Fortran, that we are using as part of a project that should allow end users to configure, launch and analyze the model from within jupyter notebooks. We are writing a python package (`pysumma`) to manage this process. We have docker instances of SUMMA, but for a number of reasons (including performance), it is desirable if we can `conda install summa` as part of the `pysumma` installation (or basically just specify a `summa` dependency for `pysumma`).

**Challenge:** SUMMA requires `gfortran` (or `gcc`) **_version 6 or greater_**. In addition, it requires a few libraries (`netcdf-fortran`, `blas`) that are consistent with the `gfortran` version being used. The default `gcc` version that is part of the `conda-forge` configuration is 4.8.*, which will not work for us. The `meta.yaml` included in this PR does not work and is more a description of how I would like it to work.

**Request:** Some suggestion about the best way to proceed would be great. I tried using `conda install -c anaconda gfortran_linux-64` (which installs `gfortran-7.2`), but there are at least two problems. For some reason, this installation does not install the `ieee_arithmetic` module (which we need), plus this `gfortran` version will likely be incompatible with the existing conda-forge `netcdf-fortran` and other libraries. In addition, it would mean that the package would rely on something outside `conda-forge`. My problems would largely disappear if a newer `gcc` version were available as part of `conda-forge`. I know there is some discussion of compiler upgrades, but I do not know how soon this is likely to happen. If a conda-wide compiler upgrade is imminent (few weeks) then it may be worth waiting, but if not, then I am looking for the best way to proceed.

### CODE REVIEW

**Comment by jhamman** [comment]
@ocefpaf - I'm wondering if you have any suggestions on how to handle the requirement for this build of `gfortran >= 6`? I'd image someone has tried to do this before but a fairly exhaustive google/github/SO search left us empty handed. 

Any thoughts from conda-forge community would be quite appreciated.

File: recipes/summa/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+build:
+  number: 0
+  skip: true # [win]
+
+requirements:
+  build:
+    - toolchain
+    - make
+    - netcdf-fortran 4.4.4
+    - gfortran >=6.0
```
Comment (jhamman) [inline recipes/summa/meta.yaml]: This is where we're currently stuck.

Comment (ocefpaf) [reply recipes/summa/meta.yaml]: @bartnijssen and @jhamman take a look at the [`scipy` recipe](https://github.com/conda-forge/scipy-feedstock/tree/master/recipe). It should have everything you need to get this one working.  I am away from my laptop, travelling to the USA for a conference, but I can give this a try once I get back.

PS: note that we are installing Fortran with the `yum_requirements.txt` file instead of using the compilers in the default channel.

File: README.md [hunk 2]
```diff
@@ -1,4 +1,9 @@
-## About
```
Comment (ocefpaf) [inline README.md]: Please do not change the staged-recipes README. The `summa` feedstock readme will be autogenerated from the recipe. All you need to do is to submit the recipe.

File: recipes/example/meta.yaml [hunk 3]
```diff
@@ -1,79 +0,0 @@
-# Note: there are many handy hints in comments in this example -- remove them when you've finalized your recipe
```
Comment (ocefpaf) [inline recipes/example/meta.yaml]: Please do not delete our example recipe. We need that for people to have something to base their recipes.

File: recipes/summa/meta.yaml (hunk 1 above)
Comment (ocefpaf) [reply recipes/summa/meta.yaml]: BTW using gfortran >=6 here, which I don't think we have either in conda-forge or in defaults, would require us to rebuild openblas, etc with that Fortran.

File: recipes/example/meta.yaml (hunk 3 above)
Comment (bartnijssen) [reply recipes/example/meta.yaml]: reverted

File: recipes/summa/meta.yaml (hunk 1 above)
Comment (bartnijssen) [reply recipes/summa/meta.yaml]: @ocefpaf : This is indeed the problem. The SUMMA code itself uses fortran features that are part of the standard, but which were not fully implemented in gfortran 4.8.* which is the default gcc version in conda-forge (and I think for a lot of the standard linux distributions). This requires rebuilding some of the libraries (netcdf, blas) to be consistent, which is the challenge. I'll take a closer look at the scipy recipe, but perhaps waiting for the compiler upgrade before adding this to conda-forge may be easier.

File: README.md (hunk 2 above)
Comment (bartnijssen) [reply README.md]: reverted

File: recipes/summa/meta.yaml (hunk 1 above)
Comment (ocefpaf) [reply recipes/summa/meta.yaml]: I double checked and `defaults` does have a modern Fortran we could use, but that would mean a deeper change in conda-forge (that must happen at sometime) to adopt those compilers.

It would require to re-build everything that depends on Fortran, starting from openblas.

TL;DR I don't think we are ready for this just yet. Conda-forge must move to conda-build 3, then adopt the modern compiler, then re-build the whole Fortran stack. It will take a while, sorry,

Ping: @mingwandroid who is maintaining that toolchain.

Comment (bartnijssen) [reply recipes/summa/meta.yaml]: Thanks for looking into this. This is precisely the feedback I needed (although not the answer I had hoped for :) ). I started putting this together thinking it was easy, but the compiler issue tripped us up. I will go a different route for now.

Comment (ocefpaf) [reply recipes/summa/meta.yaml]: > This is precisely the feedback I needed (although not the answer I had hoped for :) 

Stay tuned though. Ee do want to start using conda-build 3 and the compilers from the default channel ASAP.

**Comment by jhamman** [comment]
huzzah!

**Comment by ocefpaf** [comment]
Linux is failing with:

```
$BUILD_PREFIX/bin/x86_64-conda_cos6-linux-gnu-gfortran -p -g -Wall -ffree-line-length-none -fmax-errors=0 -fbacktrace -fcheck=bounds -c  $SRC_DIR/build/source/netcdf/netcdf_util.f90  $SRC_DIR/build/source/netcdf/def_output.f90  $SRC_DIR/build/source/netcdf/modelwrite.f90  $SRC_DIR/build/source/netcdf/read_icond.f90  $SRC_DIR/build/source/engine/conv_funcs.f90  $SRC_DIR/build/source/engine/sunGeomtry.f90  $SRC_DIR/build/source/engine/snow_utils.f90  $SRC_DIR/build/source/engine/soil_utils.f90  $SRC_DIR/build/source/engine/updatState.f90  $SRC_DIR/build/source/engine/convE2Temp.f90  $SRC_DIR/build/source/engine/allocspace.f90  $SRC_DIR/build/source/engine/checkStruc.f90  $SRC_DIR/build/source/engine/childStruc.f90  $SRC_DIR/build/source/engine/ffile_info.f90  $SRC_DIR/build/source/engine/read_attrb.f90  $SRC_DIR/build/source/engine/read_pinit.f90  $SRC_DIR/build/source/engine/pOverwrite.f90  $SRC_DIR/build/source/engine/read_param.f90  $SRC_DIR/build/source/engine/paramCheck.f90  $SRC_DIR/build/source/engine/check_icond.f90  $SRC_DIR/build/source/engine/indexState.f90  $SRC_DIR/build/source/engine/getVectorz.f90  $SRC_DIR/build/source/engine/updateVars.f90  $SRC_DIR/build/source/engine/var_derive.f90  $SRC_DIR/build/source/engine/read_force.f90  $SRC_DIR/build/source/engine/derivforce.f90  $SRC_DIR/build/source/engine/snowAlbedo.f90  $SRC_DIR/build/source/engine/canopySnow.f90  $SRC_DIR/build/source/engine/tempAdjust.f90  $SRC_DIR/build/source/engine/snwCompact.f90  $SRC_DIR/build/source/engine/layerMerge.f90  $SRC_DIR/build/source/engine/layerDivide.f90  $SRC_DIR/build/source/engine/volicePack.f90  $SRC_DIR/build/source/engine/qTimeDelay.f90  $SRC_DIR/build/source/engine/vegPhenlgy.f90  $SRC_DIR/build/source/engine/diagn_evar.f90  $SRC_DIR/build/source/engine/stomResist.f90  $SRC_DIR/build/source/engine/groundwatr.f90  $SRC_DIR/build/source/engine/vegSWavRad.f90  $SRC_DIR/build/source/engine/vegNrgFlux.f90  $SRC_DIR/build/source/engine/ssdNrgFlux.f90  $SRC_DIR/build/source/engine/vegLiqFlux.f90  $SRC_DIR/build/source/engine/snowLiqFlx.f90  $SRC_DIR/build/source/engine/soilLiqFlx.f90  $SRC_DIR/build/source/engine/computFlux.f90  $SRC_DIR/build/source/engine/computResid.f90  $SRC_DIR/build/source/engine/computJacob.f90  $SRC_DIR/build/source/engine/eval8summa.f90  $SRC_DIR/build/source/engine/summaSolve.f90  $SRC_DIR/build/source/engine/systemSolv.f90  $SRC_DIR/build/source/engine/varSubstep.f90  $SRC_DIR/build/source/engine/opSplittin.f90  $SRC_DIR/build/source/engine/coupled_em.f90  $SRC_DIR/build/source/driver/multi_driver.f90 \
	-I$PREFIX/include
/home/conda/feedstock_root/build_artifacts/summa_1566604017526/work/build/source/driver/multi_driver.f90:1107:30:

    print "(A)", spaces(1:int((70 - len_trim(summaVersion) - 9) / 2))//'Version: '   //trim(summaVersion)
                              1
Warning: Integer division truncated to constant '30' at (1) [-Winteger-division]
/home/conda/feedstock_root/build_artifacts/summa_1566604017526/work/build/source/driver/multi_driver.f90:1323:35:

  character(len=256)     :: cmessage                ! error message of the downwind routine
                                   1
Warning: Unused variable 'cmessage' declared at (1) [-Wunused-variable]
/home/conda/feedstock_root/build_artifacts/summa_1566604017526/work/build/source/driver/multi_driver.f90:1322:33:

  integer(i4b)           :: nc_err                  ! error code of nc_close
                                 1
Warning: Unused variable 'nc_err' declared at (1) [-Wunused-variable]
$BUILD_PREFIX/bin/x86_64-conda_cos6-linux-gnu-gfortran -pg -g *.o -L$PREFIX/lib -lnetcdff -L$PREFIX/lib -llapack -lblas -o summa.exe
/home/conda/feedstock_root/build_artifacts/summa_1566604017526/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: /home/conda/feedstock_root/build_artifacts/summa_1566604017526/_build_env/bin/../x86_64-conda_cos6-linux-gnu/sysroot/usr/lib/../lib/gcrt1.o: relocation R_X86_64_32S against symbol `__libc_csu_fini' can not be used when making a PIE object; recompile with -fPIC
/home/conda/feedstock_root/build_artifacts/summa_1566604017526/_build_env/bin/../lib/gcc/x86_64-conda_cos6-linux-gnu/7.3.0/../../../../x86_64-conda_cos6-linux-gnu/bin/ld: final link failed: nonrepresentable section on output
collect2: error: ld returned 1 exit status
make: *** [link] Error 1
make: Leaving directory `$SRC_DIR/build'
```

@isuruf do you have any advice for us?

**Comment by isuruf** [comment]
I think this is a bug in the glibc vendored. cc @mingwandroid 

Try patching the build system to remove `-pg`

**Comment by ocefpaf** [comment]
> I think this is a bug in the glibc vendored. cc @mingwandroid
> 
> Try patching the build system to remove `-pg`

That worked. Thanks @isuruf!

@bartnijssen and @jhamman we have a new error now that I believe is related to how `summa` links to `netcdf-fortran`:

    summa.exe -v
    summa.exe: error while loading shared libraries: libnetcdff.so.6:
    cannot open shared object file: No such file or directory

**Comment by jhamman** [comment]
Whoa! This is getting really close. Just seems to be failing on OSX. FWIW, I think we could live with linux only for now. @bartnijssen can give the final word though.

**Comment by isuruf** [comment]
You can skip gfortran 4 for osx and only build with gfortran 7 on osx.

**Comment by jhamman** [comment]
> You can skip gfortran 4 for osx and only build with gfortran 7 on osx.

@isuruf - how is this done?

**Comment by ocefpaf** [comment]
Let's get this merged with Linux only and work on OS X in the feedstock. Thanks a lot @isuruf for your help!
